catalyst/runners/runner.py

```diff
@@ -861,4 +862,82 @@ def get_callbacks(self, stage: str) -> "OrderedDict[str, Callback]":
         return callbacks
 
 
-__all__ = ["Runner", "SupervisedRunner"]
+class SelfSupervisedRunner(ISelfSupervisedRunner, Runner):
+    """Runner for experiments with contrastive model."""
+
+    def __init__(
+        self,
+        model: RunnerModel = None,
+        engine: IEngine = None,
+        input_key: str = "features",
+        target_key: str = "target",
+        loss_key: str = "loss",
+        augemention_prefix: str = "aug",
+        projection_prefix: str = "projection",
+        embedding_prefix: str = "embedding",
+    ):
+        """Init."""
+        ISelfSupervisedRunner.__init__(
+            self,
+            input_key=input_key,
+            target_key=target_key,
+            loss_key=loss_key,
+            augemention_prefix=augemention_prefix,
+            projection_prefix=projection_prefix,
+            embedding_prefix=embedding_prefix,
+        )
+        Runner.__init__(self, model=model, engine=engine)
+
+    @torch.no_grad()
+    def predict_batch(self, batch: Mapping[str, Any], **kwargs) -> Mapping[str, Any]:
+        """
+        Run model inference on specified data batch.
+
+        .. warning::
+            You should not override this method. If you need specific model
+            call, override forward() method
+
+        Args:
+            batch: dictionary with data batch from DataLoader.
+            **kwargs: additional kwargs to pass to the model
+
+        Returns:
+            Mapping[str, Any]: model output dictionary
+        """
+        batch = self._process_batch(batch)
+        batch = self.engine.sync_device(tensor_or_module=batch)
+        output = self.forward(batch, **kwargs)
+        return output
+
+    def get_callbacks(self, stage: str) -> "OrderedDict[str, Callback]":
+        """Prepares the callbacks for selected stage.
+
+        Args:
+            stage: stage name
+
+        Returns:
+            dictionary with stage callbacks
+        """
+        # I took it from supervised runner should be remade
+        callbacks = super().get_callbacks(stage=stage)
+        is_callback_exists = lambda callback_fn: any(
+            callback_isinstance(x, callback_fn) for x in callbacks.values()
+        )
+        if isinstance(self._criterion, Criterion) and not is_callback_exists(ICriterionCallback):
+            callbacks["_criterion"] = CriterionCallback(
+                input_key=f"{self._projection_prefix}_1",
+                target_key=f"{self._projection_prefix}_2",
+                metric_key=self._loss_key,
+            )
```

Comment:
by this way, you mean, that the criterion is working on 2 model's projections?
as far as we are talking about self-supervised runner - that looks current 🤔 
nevertheless, what about 2 model's embeddings or `targets` from the dataset?
how (and should we?) utilize them?

Reply:
I suppose we should utilize them for quality checks.

Reply:
could we add a `mode` args into the Runner, so we could work on `embeddings` or `projections` natively?

Reply:
hmm...
Like [this](https://github.com/catalyst-team/catalyst/pull/1278/commits/ce064ab003d508d4f7c2e53f81f33856ac203cd4)?